### PR TITLE
#136: Exasol SQLA client usage stopped working in certain scenarios

### LINKED ISSUE #136: Exasol SQLA client usage stopped working in certain scenarios
# SQLA client stopped working 

This issue was discovered by indirect usage, to be precise user(s) of the [apache-superset](https://github.com/apache/superset/) project stumbled upon this issue while using it. [apache-superset](https://github.com/apache/superset/) makes use of [SQLA](https://github.com/sqlalchemy/sqlalchemy) and [this](https://github.com/exasol/sqlalchemy-exasol) exasol dialect for SQLA. Due to the fact that the error couldn't yet narrowed down to a specific part in this library the error scenario/description does rely on [apache-superset](https://github.com/apache/superset/). Once the `root cause` is clear this issue should be updated accordingly.
 
The context in which the issue have been found can be seen [here](https://github.com/apache/superset/issues/16541#issuecomment-1121107710).

## How to reproduce
1. Setup an Apache superset installation/instance
   1. install apache superset  `python -m pip install "apache-superset[exasol]==1.5.0"`
   2. [Setup apache superset](https://superset.apache.org/docs/intro)
2. Add an exasol DB
    1. [How to add a DB](https://superset.apache.org/docs/databases/db-connection-ui) 
    2. [Exasol DB connection string(s)](https://superset.apache.org/docs/databases/exasol)
3. Try to add a Dataset from the Exasol DB
   1. Select the Exasol DB
   3. Select Schema
   4. Select Table/View

    image

### Expected Behavior 
:heavy_check_mark: On step `3.iii` all available Tables/Views are shown and can be selected.

### Actual Behavior
:boom: No Tables/Views are shown.

### PR DESCRIPTION
**Pull request by Nicoretti**
Fixes #136

- Add regression test for Github issue #136
- Improve translate map regression
- Remove unnecessary parameters from function
- Fix typo's in test names

### CODE REVIEW

File: sqlalchemy_exasol/base.py [hunk 1]
```diff
@@ -510,228 +487,131 @@ def _get_schema_for_input_or_current(self, connection, schema):
         return self.denormalize_name(schema)
 
     def _get_schema_for_input(self, connection, schema):
-        schema = self.denormalize_name(schema or self._get_schema_from_url(connection, schema))
-        return schema
+        return self.denormalize_name(
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py:490]: you really like to do some computation directly in the return. I usually, try to avoid this at all costs, because it makes debugging more complicated.

File: sqlalchemy_exasol/pyodbc.py [hunk 2]
```diff
@@ -164,5 +167,139 @@ def is_disconnect(self, e, connection, cursor):
...
+
+        tables = self._tables_for_schema(connection, schema=schema, table_name=table_name)
+        if len(tables) != 1:
+            return []
+
+        # get_columns_sql originally returned all columns of all tables if table_name is None,
+        # we follow this behavior here for compatibility. However, the documentation for Dialects
+        # does not mention this behavior:
+        # https://docs.sqlalchemy.org/en/13/core/internals.html#sqlalchemy.engine.interfaces.Dialect
+        quoted_schema_string = self.quote_string_value(tables[0].table_schem)
```
Comment (tkilias) [inline sqlalchemy_exasol/pyodbc.py:249]: we might want to check in another PR, if we can replace the quote_string_value with parameterized queries and parameters in connection.execute(). Please, create a ticket for that. We have also some more places with that.

File: sqlalchemy_exasol/base.py [hunk 3]
```diff
@@ -895,13 +744,7 @@ def fkey_rec():
             }
 
         fkeys = util.defaultdict(fkey_rec)
-        odbc_connection = self.getODBCConnection(connection)
-        if odbc_connection is not None and not self.use_sql_fallback(**kw):
-            constraints = self._get_foreign_keys_odbc(connection, odbc_connection, table_name=table_name,
-                                                      schema=schema_int, **kw)
-        else:
-            constraints = self._get_foreign_keys_sql(connection, table_name=table_name, schema=schema_int, **kw)
-        table_name = self.denormalize_name(table_name)
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py:904]: hm, why did we have this here, is now the question, I can't see a reason, maybe a left over

File: sqlalchemy_exasol/pyodbc.py [hunk 4]
```diff
@@ -164,5 +167,139 @@ def is_disconnect(self, e, connection, cursor):
...
 
+    @staticmethod
+    def _is_sql_fallback_requested(**kwargs):
+        is_fallback_requested = kwargs.get("use_sql_fallback", False)
+        if is_fallback_requested:
+            logger.warning("Using sql fallback instead of odbc functions")
+        return is_fallback_requested
+
+    @reflection.cache
+    def _tables_for_schema(self, connection, schema, table_type=None, table_name=None):
```
Comment (tkilias) [inline sqlalchemy_exasol/pyodbc.py:178]: _get_tables_for_schema or the name intended and if why?

File: sqlalchemy_exasol/pyodbc.py [hunk 5]
```diff
@@ -164,5 +167,139 @@ def is_disconnect(self, e, connection, cursor):
...
+        schema = self._get_schema_for_input_or_current(connection, schema)
+        table_name = self.denormalize_name(table_name)
+        conn = connection.engine.raw_connection()
```
Comment (tkilias) [inline sqlalchemy_exasol/pyodbc.py:181]: is this engine attribute garantueed?

**Review by tkilias — CHANGES_REQUESTED** [review]

File: sqlalchemy_exasol/pyodbc.py (hunk 5 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/pyodbc.py:181]: From what I have seen in the docs and the code, `yes`.

File: sqlalchemy_exasol/pyodbc.py (hunk 4 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/pyodbc.py:178]: was intended that way, don't think it adds beneficial information to this non public method.

File: sqlalchemy_exasol/base.py (hunk 3 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/base.py:904]: I don't know the IDE just pointed out it that it isn't used

File: sqlalchemy_exasol/base.py (hunk 1 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/base.py:490]: I can see your point, personally I would add it when debugging it if necessary or setup a watch expression. If you think though we should add it I am fine with it too.

Comment (Nicoretti) [reply sqlalchemy_exasol/base.py:490]: Reverted ;)

File: sqlalchemy_exasol/pyodbc.py [hunk 6]
```diff
@@ -164,5 +167,139 @@ def is_disconnect(self, e, connection, cursor):
...
+                constraint_name = self.normalize_name(constraint)
+        return {'constrained_columns': pkeys, 'name': constraint_name}
+
+    @reflection.cache
+    def _get_foreign_keys(self, connection, table_name, schema=None, **kw):
+        if self._is_sql_fallback_requested(**kw):
+            return super()._get_foreign_keys(connection, table_name, schema, **kw)
+
+        # Need to use a workaround, because SQLForeignKeys functions doesn't work for an unknown reason
+        tables = self._tables_for_schema(
```
Comment (tkilias) [inline sqlalchemy_exasol/pyodbc.py:282]: Needs to be checked in another PR

File: sqlalchemy_exasol/pyodbc.py (hunk 2 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/pyodbc.py:249]: See [issue](https://github.com/exasol/sqlalchemy-exasol/issues/138)

File: sqlalchemy_exasol/pyodbc.py [hunk 7]
```diff
@@ -164,5 +167,139 @@ def is_disconnect(self, e, connection, cursor):
...
+            return []
+
+        # get_columns_sql originally returned all columns of all tables if table_name is None,
+        # we follow this behavior here for compatibility. However, the documentation for Dialects
+        # does not mention this behavior:
+        # https://docs.sqlalchemy.org/en/13/core/internals.html#sqlalchemy.engine.interfaces.Dialect
+        quoted_schema_string = self.quote_string_value(tables[0].table_schem)
+        quoted_table_string = self.quote_string_value(tables[0].table_name)
+        sql_statement = "/*snapshot execution*/ {query}".format(query=self.get_column_sql_query_str())
+        sql_statement = sql_statement.format(schema=quoted_schema_string, table=quoted_table_string)
```
Comment (tkilias) [inline sqlalchemy_exasol/pyodbc.py:252]: We need to check if we can replace this query with a odbc metadata function.

File: sqlalchemy_exasol/pyodbc.py (hunk 6 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/pyodbc.py:282]: See https://github.com/exasol/sqlalchemy-exasol/issues/139

File: sqlalchemy_exasol/pyodbc.py (hunk 7 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/pyodbc.py:252]: See https://github.com/exasol/sqlalchemy-exasol/issues/139

**Review by tkilias — APPROVED** [review]
